src/analysis/processing/qgsalgorithmrandomraster.cpp

```diff
@@ -24,34 +24,184 @@
 ///@cond PRIVATE
 
 //
-// QgsRandomRasterAlgorithm
+// QgsRandomRasterAlgorithmBase
 //
-QString QgsRandomRasterAlgorithm::name() const
+QString QgsRandomRasterAlgorithmBase::group() const
 {
-  return QStringLiteral( "createrandomrasterlayer" );
+  return QObject::tr( "Raster creation" );
 }
 
-QString QgsRandomRasterAlgorithm::group() const
+QString QgsRandomRasterAlgorithmBase::groupId() const
 {
-  return QObject::tr( "Raster creation" );
+  return QStringLiteral( "rastercreation" );
 }
 
-QString QgsRandomRasterAlgorithm::groupId() const
+void QgsRandomRasterAlgorithmBase::initAlgorithm( const QVariantMap & )
 {
-  return QStringLiteral( "rastercreation" );
+  addParameter( new QgsProcessingParameterExtent( QStringLiteral( "EXTENT" ), QObject::tr( "Desired extent" ) ) );
+  addParameter( new QgsProcessingParameterCrs( QStringLiteral( "TARGET_CRS" ), QObject::tr( "Target CRS" ), QStringLiteral( "ProjectCrs" ) ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "PIXEL_SIZE" ), QObject::tr( "Pixel size" ),
+                QgsProcessingParameterNumber::Double, 1, false, 0.01 ) );
+
+  //add specific parameters
+  addAlgorithmParams();
+
+  addParameter( new QgsProcessingParameterRasterDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Output raster" ) ) );
+}
+
+bool QgsRandomRasterAlgorithmBase::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+  mCrs = parameterAsCrs( parameters, QStringLiteral( "TARGET_CRS" ), context );
+  mExtent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context, mCrs );
+  mPixelSize = parameterAsDouble( parameters, QStringLiteral( "PIXEL_SIZE" ), context );
+
+  int typeId = parameterAsInt( parameters, QStringLiteral( "OUTPUT_TYPE" ), context );
+  //prepare specific parameters
+  mRasterDataType = getRasterDataType( typeId );
+  prepareRandomParameters( parameters, context );
```

Comment:
I think this should be moved to processAlgorithm, to ensure it's done in the same thread as the distributions are used from.

Reply:
Ok, np! The `prepareAlgorithm()` method is a mystery to me - on the one hand it is used to prepare the algorithm execution, on the other hand it is barely used to do so in most algorithms (except the ones written by me so far^^). I know that `processAlgorithm()` runs in a separate thread - but why is `prepareAlgorithm()` even needed if all parameter reading is done within `processAlgorithm()`?

Reply:
Haha - on the second read - ignore my question. I understand your concern: Thread safety and distribution objects. :smile: